src/rpc/notifications.cpp

```diff
@@ -0,0 +1,381 @@
+// Copyright (c) 2017 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "chain.h"
+#include "init.h"
+#include "primitives/block.h"
+#include "primitives/transaction.h"
+#include "rpc/server.h"
+#include "sync.h"
+#include "util.h"
+#include "utilstrencodings.h"
+#include "validationinterface.h"
+
+#include <univalue.h>
+
+#include <boost/assign/list_of.hpp>
+#include <boost/foreach.hpp>
+#include <boost/range/adaptors.hpp>
+
+#include <deque>
+#include <exception>
+#include <stdint.h>
+
+static const char* MSG_HASHBLOCK = "hashblock";
+static const char* MSG_HASHTX = "hashtx";
+
+/* keep the max queue size large becase we don't
+   auto-register for notification on startup */
+static const size_t MAX_QUEUE_SIZE = 1024 * 1024;
+static const int DEFAULT_POLL_TIMEOUT = 30;
+
+enum class NotificationType {
+    Unknown,
+    Block,
+    Tx
+};
+
+typedef std::pair<size_t, size_t> queueRange_t;
+typedef std::string clientUUID_t;
+
+// class that represents a notification
+class NotificationEntry
+{
+public:
+    NotificationType m_type;
+    int32_t m_sequence_number;
+    UniValue m_notification;
+};
+
+class NotificationQueue
+{
+public:
+    std::deque<NotificationEntry> m_queue;
+    std::map<NotificationType, int32_t> m_map_sequence_numbers;
```

Comment:
I generally prefer 64 bit ints here - sure, its unlikely you'd overflow 32 bits, but if you're online for 3 or 4 years you may start getting closeish.